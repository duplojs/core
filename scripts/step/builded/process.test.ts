import { Process } from "@scripts/duplose/process";
import { ProcessStep, type ProcessStepParams } from "../process";
import { BuildedProcessStep } from "./process";
import { Duplo } from "@scripts/duplo";
import { makeFloor } from "@scripts/floor";
import { readFile } from "fs/promises";
import { resolve } from "path";
import { duploTest } from "@test/utils/duploTest";

describe("BuildedProcessStep", () => {
	const process = new Process("test");
	process.instance = duploTest;
	process.setOptions({
		toto: 1,
		test: "",
	});
	process.setInput(22);

	it("merge object options", () => {
		const params: ProcessStepParams = {
			options: {
				toto: 2,
				test1: "&",
			},
		};

		const step = new ProcessStep(process, params);

		const buildedProcessStep = new BuildedProcessStep(duploTest, step);

		expect(buildedProcessStep.params.options).toStrictEqual({
			toto: 2,
			test: "",
			test1: "&",
		});

		expect(buildedProcessStep.params.input?.(makeFloor().pickup)).toBe(22);
	});

	it("merge function options", () => {
		const params: ProcessStepParams = {
			options: () => ({
				toto: 2,
				test1: "&",
			}),
		};

		const step = new ProcessStep(process, params);

		const buildedProcessStep = new BuildedProcessStep(duploTest, step);

		expect(
			typeof buildedProcessStep.params.options === "function"
				? buildedProcessStep.params.options(makeFloor().pickup)
				: undefined,
		).toStrictEqual({
			toto: 2,
			test: "",
			test1: "&",
		});
	});

	it("toString options: function, pickup", async() => {
		const params: ProcessStepParams = {
			pickup: ["rr", "tt"],
			options: () => ({
				toto: 2,
				test1: "&",
			}),
		};

		const step = new ProcessStep(process, params);

		const buildedProcessStep = new BuildedProcessStep(duploTest, step);

		expect(buildedProcessStep.toString(1)).toBe(
			await readFile(resolve(import.meta.dirname, "__data__/process1.txt"), "utf-8"),
		);
	});

	it("toString options: object, input, skip", async() => {
		const params: ProcessStepParams = {
			input: () => undefined,
			options: {
				toto: 2,
				test1: "&",
			},
			skip: () => true,
		};

		const step = new ProcessStep(process, params);

		const buildedProcessStep = new BuildedProcessStep(duploTest, step);

		expect(buildedProcessStep.toString(1)).toBe(
			await readFile(resolve(import.meta.dirname, "__data__/process2.txt"), "utf-8"),
		);
	});
});
