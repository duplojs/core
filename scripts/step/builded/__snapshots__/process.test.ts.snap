// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BuildedProcessStep > toString options: function, pickup 1`] = `
"
/* beforeProcessStep(index: 1) */
/* end_block */


				result = this.steps[1].processFunction(
					request,
					this.steps[1].params.options(floor.pickup),
					this.steps[1].params.input(floor.pickup)
				);

				
/* beforeTreatResultProcessStep(index: 1) */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* beforeIndexingResultProcessStep(index: 1) */
/* end_block */


				floor.drop("rr", result["rr"]);
floor.drop("tt", result["tt"]);
				

/* afterProcessStep(index: 1) */
/* end_block */
"
`;

exports[`BuildedProcessStep > toString options: object, input, skip 1`] = `
"
/* beforeProcessStep(index: 1) */
/* end_block */


		if(!this.steps[1].params.skip(floor.pickup)){
			
				result = this.steps[1].processFunction(
					request,
					this.steps[1].params.options,
					this.steps[1].params.input(floor.pickup)
				);

				
/* beforeTreatResultProcessStep(index: 1) */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* beforeIndexingResultProcessStep(index: 1) */
/* end_block */


				
				
		}
		

/* afterProcessStep(index: 1) */
/* end_block */
"
`;
